File: code_samples/go/error_handling_error_propagation.go
package main

import (
    "fmt"
)

// Error Handling: Error Propagation
// AI/ML Training Sample

type ErrorPropagation struct {
    Data string
}

func NewErrorPropagation() *ErrorPropagation {
    return &ErrorPropagation{
        Data: "",
    }
}

func (s *ErrorPropagation) Process(input string) {
    s.Data = input
}

func (s *ErrorPropagation) Validate() bool {
    return len(s.Data) > 0
}

func (s *ErrorPropagation) GetData() string {
    return s.Data
}

func main() {
    instance := NewErrorPropagation()
    instance.Process("example")
    fmt.Printf("Data: %s\n", instance.GetData())
    fmt.Printf("Valid: %v\n", instance.Validate())
}

File: code_samples/haskell/error_handling_error_propagation.hs
-- Error Handling: Error Propagation
-- AI/ML Training Sample

module ErrorPropagation where

data ErrorPropagation = ErrorPropagation {
    getData :: String
} deriving (Show, Eq)

process :: ErrorPropagation -> String -> ErrorPropagation
process obj input = obj { getData = input }

validate :: ErrorPropagation -> Bool
validate obj = not (null (getData obj))

main :: IO ()
main = do
    let instance = ErrorPropagation { getData = "" }
    let updated = process instance "example"
    putStrLn $ "Data: " ++ getData updated
    putStrLn $ "Valid: " ++ show (validate updated)
